### ISSUE TITLE
Some animations load improperly

### ISSUE THREAD

**Issue by z64me** [Dragorn421/zelda64-import-blender#1]
The left screenshot is the latest importer, whereas the right is the same ZOBJ imported with an older version.
Settings were left as default for each import:

![zelda64-import-blender](https://user-images.githubusercontent.com/45853441/89721477-eb632600-d9ab-11ea-8315-4cc8d2a3facf.jpeg)

Here is the ZOBJ I'm importing:
[zelda_skelton_new_generic.zip](https://github.com/Dragorn421/zelda64-import-blender/files/5046569/zelda_skelton_new_generic.zip)

Here is the version of the importer that yields the correct result:
[io_import_z64_latest.zip](https://github.com/Dragorn421/zelda64-import-blender/files/5046571/io_import_z64_latest.zip)

**Comment by Dragorn421** [comment]
Any idea where you got that importer version from, and when? Or even who last touched it?
I'm pretty sure it's nowhere in the commit history, but I made sure to include all releases by SoulOfDeity and Nokaubure, or so I thought.

**Comment by z64me** [comment]
It also works upon reverting to this version (`Zelda64 Importer v4 by SoulofDeity`):

https://github.com/Dragorn421/zelda64-import-blender/blob/9596d5ddcbcfcae93201878545b710ac67df51e4/io_import_z64.py

**Comment by Dragorn421** [comment]
Here is what I got with the [latest commit](https://github.com/Dragorn421/zelda64-import-blender/tree/789c69f45c8be47d09efc931e230979b833c5389) with default settings, **after setting the active action to `Anim1_7`** (and switching frame to update the bone rotations):

![image](https://user-images.githubusercontent.com/5306619/89741171-5afe1180-da8f-11ea-8b94-a8d900212693.png)

Here is first frame of `Anim1_7` with Texture Clamp/Mirror and Prim/Env Color options ON (they default to OFF in recent versions)

![image](https://user-images.githubusercontent.com/5306619/89741315-79b0d800-da90-11ea-8a8b-9e0eaa5c794f.png)

The active action after the import is done is always the last imported action (in this case the 12th `Anim12_20`) which I guess here may either be a junk unused animation in the zobj, or data wrongly detected as animation (I doubt that though, because most of the rotations make sense, only the skull is messed up)

The change of behavior between the old script and the new one is that the newer versions import all animations at once, while old versions only imported one animations (the first one detected in the zobj by default).

TLDR and takeaway: Open the console while importing, notice there are several animations, try them all. Some are messed up, some are ok

**Comment by z64me** [comment]
Now testing with the latest version again, the Blender console reports it has imported 12/12 animations. Only one animation appears correct (the `anim1_7` you cited). The other eleven are broken. Here is a video showcasing all twelve animations displaying properly in Hylian Toolbox: https://youtu.be/aWCfz3rZBEg

The latest Utility of Time ([available here](https://code.google.com/archive/p/uot/)) is also capable of displaying these animations properly.

Because Hylian Toolbox and Utility of Time do not support branch instructions, a modified ZOBJ must be opened in them instead. Here is the one I'm using, which exhibits the same problems with the plugin as the one I posted earlier: [zelda_skelton_new_unde.zip](https://github.com/Dragorn421/zelda64-import-blender/files/5048671/zelda_skelton_new_unde.zip)

Edit to add that the animation offsets and frame counts detected by the plugin match those displayed in the right pane in the Hylian Toolbox video:
```
INFO:z64import.F3DZEX.locateAnimations:           Anims found at 00014008 Frames: 7
INFO:z64import.F3DZEX.locateAnimations:           Anims found at 00015660 Frames: 50
INFO:z64import.F3DZEX.locateAnimations:           Anims found at 00015B90 Frames: 13
INFO:z64import.F3DZEX.locateAnimations:           Anims found at 000162E8 Frames: 20
INFO:z64import.F3DZEX.locateAnimations:           Anims found at 000168E0 Frames: 14
INFO:z64import.F3DZEX.locateAnimations:           Anims found at 00017678 Frames: 39
INFO:z64import.F3DZEX.locateAnimations:           Anims found at 00019258 Frames: 60
INFO:z64import.F3DZEX.locateAnimations:           Anims found at 0001A600 Frames: 47
INFO:z64import.F3DZEX.locateAnimations:           Anims found at 0001BC58 Frames: 50
INFO:z64import.F3DZEX.locateAnimations:           Anims found at 0001C388 Frames: 16
INFO:z64import.F3DZEX.locateAnimations:           Anims found at 0001C938 Frames: 13
INFO:z64import.F3DZEX.locateAnimations:           Anims found at 0001D070 Frames: 20
INFO:z64import.F3DZEX.locateAnimations:           Total Anims                         : 12
```

**Comment by Dragorn421** [comment]
I confirmed that at least animation 2 (offset 0x15660) in zelda_skelton_new_unde.zobj is already messed up in an [early version (Nokaubure's)](https://github.com/Dragorn421/zelda64-import-blender/tree/59f051bad648a205867dc423bccac54db462deb0).
The same issue shows in the [latest version](https://github.com/Dragorn421/zelda64-import-blender/tree/789c69f45c8be47d09efc931e230979b833c5389)
Just looking at it it seems that at least limb_03 is using very weird rotations values, making the upper half of the body rotate around y (several turns!) unlike in HT, eg at frame 29 Euler XYZ is `8 -4 -19` in HT and `-65 -6 -83` in Blender. Will investigate more

**Comment by LordFrijole** [comment]
Hello, how do you import animations and view them in blender?

**Comment by Dragorn421** [comment]
Hi, could you come over on the Hylian Modding discord, it would make it easier to communicate and there are more people there

You can find a link to the discord on the Hylian Modding website: https://hylianmodding.com/ (scroll down)

**Comment by LordFrijole** [comment]
Hello. I'm having a little bit of difficulty browsing the discord on where
to comment. I did figure out how animations work with this script. However
I wanted to let you know of a discovery i've made.

I'm not sure if you've already found this out, but the old import script
that is referenced in that commit on github is actually capable of
importing Link's animations.
I have gamekeep renamed to segment04.zdata and the link animation zdata
renamed to segment07.zdata loaded on the same folder of the link boy zobj
when importing. It only works to import one animation at a time, but it
works, this script version can load all of link's animations.

I also found that the older script sometimes loads animations more
accurately positioning wise, but the amount of movement isn't correct.

Specifically with importing the eyegore zobj from majora's mask (the id
being: 017AD000 - 017C01F0). If the eyegore is loaded using the newest
script, the legs in particular are positioned inaccurately. When loaded
using the older script, every limb is positioned accurately for each frame
in the animation, however the movement is erratic with an up and down
movement for each frame (This is with animation 1, when importing with the
old script.)

I found that a fix for this is to import the eyegore with both scripts: By
loading the action from the eyegore imported using the older script into
the eyegore that was imported with the newer script, the animation will
play properly. I notice that the newer script loads zobjs in a larger
scale, whilst the older one is much smaller. The older script seems to load
the animation meant for a larger scale which could explain the erratic
movement.

I'm not sure if that makes sense the way I explained it but this is what I
found to work. Thank you for your work on this amazing script, i'm having
fun with it!

**Comment by Dragorn421** [comment]
"older script" in the context of this message: https://cdn.discordapp.com/attachments/451783162859749386/905955051997126696/io_import_z64_older_script.py

https://discord.com/channels/388361645073629187/451783162859749386/905954013189644318 :

> The older one is capable of loading most actor animations properly. The newer script the animations can have certain limbs at incorrect angles/positions.
> This older script is also able to load link's animations
> Here are blend examples showing the results of importing with each script:
> https://cdn.discordapp.com/attachments/451783162859749386/905956973516845156/eyegore_newer_script.blend
> https://cdn.discordapp.com/attachments/451783162859749386/905956992538001498/eyegore_older_script.blend
> The animation of the eyegore in the blend file using the new script shows the eyegore’s current selected animation to be incorrect with its legs rotated improperly.

TODO:
1) figure out what version the "older script" is (probably Nokaubure's?)
2) fix this jank animation code for good
3) bring back link animations. It isn't that complicated

EDIT:
that "older script" looks like en early version of [Nokaubure's changes](https://github.com/Dragorn421/zelda64-import-blender/blob/59f051bad648a205867dc423bccac54db462deb0/io_import_z64.py)
it mainly has type 1 meshes and, weirdly, some commented out code ("Please do not do that") that makes it in in what I know as the Noka-edited importer (the revision linked above).
As far as animations go, it doesn't look different from the SoD's v4 or Nokaubure's changes revisions

**Comment by Dragorn421** [comment]
\+ in most of her animations, Saria's left leg ? (object 188)
